Control-based Router and Transaction Estimator

SLMT opened the pull request:
This PR adds an API to build estimation for transactions and alters the flow of stored procedure calls to send estimation to all DB servers. I also added a router based on the estimation and a controller to control the parameters of the router.

This is still **work-in-progress**. Here is the TODO list:

- Router & Estimator
  - [x] Transaction Estimation
  - [x] Estimator
  - [x] Path to send estimations
  - [x] Router
  - [x] Test
- Centralized Control Actuator
  - [x] PID Controller
  - [x] Centralized Actuator
  - [x] Control Parameter Update Transactions
  - [x] Test

SLMT committed:
Add TransactionEstimation and move a property

SLMT committed:
Change the point of deciding transaction numbers

SLMT committed:
Add HERMES-CONTROL system type

SLMT committed:
Add control-based router

SLMT committed:
Merge remote-tracking branch 'origin/research/hermes-control' into research/hermes-control-pid

SLMT committed:
Build up a path for sending estimation from the sequencer

SLMT committed:
Add ControlBasedRouter and fix a few bugs

- Fix the problem that SpCallPreprocessor is null when ENABLE_COLLECTING_DATA is false
- Add a counter in the router for debugging
- Fix the bug that node 0 cannot get any txn during ties

SLMT committed:
Add control actuator for control-based routing

SLMT committed:
Create a package for controllers

SLMT committed:
Add a special transaction to update control parameters

SLMT committed:
Fix a bug that ControlStoredProcedureFactory is not used

xuan19970825 committed:
Fix a few bugs

SLMT committed:
Merge remote-tracking branch 'origin/research/hermes-control' into research/hermes-control-pid

pin-yu commented on src/main/java/org/elasql/perf/tpart/TPartPerformanceManager.java:
```diff
@@ -4,57 +4,66 @@
+			metricWarehouse = new TpartMetricWarehouse();
+			Elasql.taskMgr().runTask(metricWarehouse);
+			
+			// The sequencer maintains a feature collector and a warehouse
```
Feature collector is removed. This comment should be revised.

pin-yu commented on src/main/java/org/elasql/perf/tpart/TPartPerformanceManager.java:
```diff
@@ -4,57 +4,66 @@
+					graph, isBatching, metricWarehouse, estimator);
+			Elasql.taskMgr().runTask(spCallPreprocessor);
+			
+			// Hermes-Control has a control aucuator
```
typo `aucuator` -> `actuator`

pin-yu commented on src/main/java/org/elasql/perf/tpart/control/ControlParamUpdateParamHelper.java:
```diff
@@ -0,0 +1,49 @@
+
+public class ControlParamUpdateParamHelper extends StoredProcedureParamHelper {
+	
+	private double[] alpha;
```
We could add some comments to clarify what alpha, beta, gamma are

pin-yu commented on src/main/java/org/elasql/perf/tpart/control/RoutingControlActuator.java:
```diff
@@ -0,0 +1,118 @@
+	
+	@Override
+	public void run() {
+		Thread.currentThread().setName("routing-control-aucuator");
```
typo `aucuator` -> actuator

pin-yu commented on src/main/java/org/elasql/perf/tpart/metric/TpartMetricWarehouse.java:
```diff
@@ -55,22 +55,38 @@ private synchronized void recordMetric(TPartSystemMetrics metrics) {
+		if (load == null)
+			return 0.0;
+		else
+		return load;
```
wrong indentation

pin-yu reviewed (changes requested):
Good code quality and readability.
There are just a few typos.
Let's fix it and go

SLMT committed:
Merge branch 'research/hermes-control' into research/hermes-control-pid

SLMT committed:
A few updates

- Update the scale of time delta
- Make the actuator wait for a delay before starting
- Make the actuator update faster
- Make the metric collector collect metrics more fequently

SLMT committed:
Fix a few typos

SLMT commented:
@pin-yu I fixed the issues. Please check again.

SLMT committed:
Fix a typo

SLMT committed:
Update and rename `StupidEstimator`

SLMT committed:
Fix a but that Calvin cannot work

pin-yu reviewed (approved):
Looks better.
Let's merge it

xuan19970825 committed:
Smoothing reference values

SLMT committed:
Add feature: "Read Data Distribution"

xuan19970825 committed:
Format cpu load and system load

SLMT committed:
Add a method to get an average of CPU loads over time

SLMT committed:
Merge remote-tracking branch 'origin/research/hermes-control-pid' into research/hermes-control-pid

SLMT committed:
Exclude fully replicated data from data distribution

SLMT committed:
Revert 'Smoothing reference values'

This seems not needed after we updated the way of retrieving observation.
